### PR TITLE
123 dialog

### PR DESCRIPTION
**Pull request by MaksGMaks**
Dialogs work, maybe i can add text to login and registration dialogs bcs it's not visible which line is login and so on

### CODE REVIEW

File: _client/src/Dialogs/DialogManager.hpp [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+#include <QObject>
+#include <QMessageBox>
+
+#include "LoginDialog.hpp"
+#include "RegistrationDialog.hpp"
+#include "Network/ApiManager.hpp"
+
+/**
+ * @class DialogManager
+ * Controls all dialogs in MessageBoxes
```
Comment (Sage-Cat) [inline _client/src/Dialogs/DialogManager.hpp]: Refactor comment to make more sense

File: _client/src/UiManager.cpp [hunk 2]
```diff
@@ -36,15 +38,15 @@ void UiManager::init()
     SPDLOG_TRACE("UiManager::init");
 
     initMainWindow();
```
Comment (Sage-Cat) [inline _client/src/UiManager.cpp]: We don't need methods initMainWindow and initDialogManager
Just do initialization here simple as `m_mainWindow = new ...`

File: _client/src/Dialogs/DialogManager.hpp [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+{
+    Q_OBJECT
+
+protected:
+public:
+    /**
+     * @brief Constructor and destructor of DialogManager, which control other objects
+     */
+    DialogManager(ApiManager *apiManager);
+    virtual ~DialogManager();
```
Comment (Sage-Cat) [inline _client/src/Dialogs/DialogManager.hpp]: virtual is not necessary

File: _client/src/Dialogs/DialogManager.hpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+#include "LoginDialog.hpp"
+#include "RegistrationDialog.hpp"
+#include "Network/ApiManager.hpp"
+
+/**
+ * @class DialogManager
+ * Controls all dialogs in MessageBoxes
+ */
+class DialogManager : public QObject
```
Comment (Sage-Cat) [inline _client/src/Dialogs/DialogManager.hpp]: make it final https://en.cppreference.com/w/cpp/language/final

File: _client/src/Dialogs/DialogManager.hpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    void showRegistrationDialog();
+
+private:
+    /**
+     * @brief initialization of dialogs
+     */
+    void initDialogs();
+
+    /**
+     * @brief setup connections for ApiManager
```
Comment (Sage-Cat) [inline _client/src/Dialogs/DialogManager.hpp]: Not for ApiManager, but between ApiManager and dialogs

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _client/src/Dialogs/DialogManager.hpp (hunk 3 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/DialogManager.hpp]: virtual is deleted

File: _client/src/Dialogs/DialogManager.hpp (hunk 5 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/DialogManager.hpp]: changed

File: _client/src/Dialogs/DialogManager.hpp (hunk 4 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/DialogManager.hpp]: done

File: _client/src/UiManager.cpp (hunk 2 above)
Comment (MaksGMaks) [reply _client/src/UiManager.cpp]: done

File: _client/src/Dialogs/DialogManager.hpp (hunk 1 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/DialogManager.hpp]: done

**Review by Sage-Cat — DISMISSED** [review]

**Comment by Sage-Cat** [comment]
> Dialogs work, maybe i can add text to login and registration dialogs bcs it's not visible which line is login and so on

@MaksGMaks yes, it's a good idea. Do it.
And same labels is needed for registration.

File: _client/src/Dialogs/LoginDialog.cpp [hunk 6]
```diff
@@ -41,8 +41,29 @@ void LoginDialog::setupLayout()
     SPDLOG_TRACE("LoginDialog::setupLayout");
 
     auto *mainLayout = new QVBoxLayout(this);
-    mainLayout->addWidget(m_usernameField);
-    mainLayout->addWidget(m_passwordField);
+
+    ///> Layout for labels
+    std::vector<QWidget *> labels;
+    labels.push_back(new QLabel(tr("Username")));
```
Comment (Sage-Cat) [inline _client/src/Dialogs/LoginDialog.cpp]: m_usernameLable

File: _client/src/Dialogs/LoginDialog.cpp [hunk 7]
```diff
@@ -41,8 +41,29 @@ void LoginDialog::setupLayout()
...
+    labels.push_back(new QLabel(tr("Password")));
+
+    auto *labelLayout = Utils::createLayout(labels);
+
+    ///> Layout for lines
+    std::vector<QWidget *> lines;
+    lines.push_back(m_usernameField);
+    lines.push_back(m_passwordField);
+
+    auto *lineLayout = Utils::createLayout(lines);
```
Comment (Sage-Cat) [inline _client/src/Dialogs/LoginDialog.cpp]: createLayout({m_usernameField, m_passwordField})

File: _client/src/Utils.cpp [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#include "Utils.hpp"
+
+QVBoxLayout *Utils::createLayout(std::vector<QWidget *> widgets)
```
Comment (Sage-Cat) [inline _client/src/Utils.cpp]: createVBoxLayout + second `this` param to set parent (should be optional, nullptr by default)

File: _client/src/Dialogs/LoginDialog.cpp [hunk 9]
```diff
@@ -41,8 +41,29 @@ void LoginDialog::setupLayout()
...
+
+    ///> Layout for labels and lines
+    auto *inputLayout = new QHBoxLayout(this);
```
Comment (Sage-Cat) [inline _client/src/Dialogs/LoginDialog.cpp]: create also func `createHBoxLayout`

File: _client/src/Dialogs/RegistrationDialog.cpp [hunk 10]
```diff
@@ -52,9 +52,31 @@ void RegistrationDialog::setupLayout()
...
+
+    auto *labelLayout = Utils::createLayout(labels);
+
+    ///> Layout for lines
+    std::vector<QWidget *> lines;
+    lines.push_back(m_usernameField);
+    lines.push_back(m_passwordField);
+    lines.push_back(m_confirmPasswordField);
+
+    auto *lineLayout = Utils::createLayout(lines);
```
Comment (Sage-Cat) [inline _client/src/Dialogs/RegistrationDialog.cpp]: same here as in LoginDialog

File: _client/src/Dialogs/RegistrationDialog.cpp [hunk 11]
```diff
@@ -52,9 +52,31 @@ void RegistrationDialog::setupLayout()
...
-    mainLayout->addWidget(m_passwordField);
-    mainLayout->addWidget(m_confirmPasswordField);
+
+    ///> Layout for labels
+    std::vector<QWidget *> labels;
+    labels.push_back(new QLabel(tr("Username")));
+    labels.push_back(new QLabel(tr("Password")));
+    labels.push_back(new QLabel(tr("Confirm password")));
+
+    auto *labelLayout = Utils::createLayout(labels);
```
Comment (Sage-Cat) [inline _client/src/Dialogs/RegistrationDialog.cpp]: You need to remember, that Qt handles dynamic object memory allocation only for child QObjects.
Here you just create labelLayout, so you need either manage deletion of the object by yourself (`delete labelLayout` in DialogManager destructor) or use `this` pointer when creating object, so DialogManager will manage memory for you.

File: _client/src/UiManager.cpp [hunk 12]
```diff
@@ -19,11 +22,10 @@ UiManager::UiManager(ApiManager *apiClient, QObject *parent) noexcept
     // Initialize all necessary elements
     init();
     setupApiConnections();
-    setupDialogsConnections();
     setupMVVMConnections();
 
     // Before initialization
-    m_mainWindow->hide();
+    m_mainWindow->setEnabled(false);
```
Comment (Sage-Cat) [inline _client/src/UiManager.cpp]: If you want to keep this change, then you need to implement logic of `setEnable(true)` after `LoginDialog` gets `LoginSuccess`

**Review by Sage-Cat — CHANGES_REQUESTED** [review]

File: _client/src/Dialogs/LoginDialog.cpp (hunk 9 above)
Comment (Sage-Cat) [reply _client/src/Dialogs/LoginDialog.cpp]: you have duplicated code in LoginDialog and in RegisterDialog

File: _client/src/Dialogs/LoginDialog.cpp (hunk 6 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/LoginDialog.cpp]: done

File: _client/src/Dialogs/LoginDialog.cpp (hunk 7 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/LoginDialog.cpp]: done

File: _client/src/Utils.cpp (hunk 8 above)
Comment (MaksGMaks) [reply _client/src/Utils.cpp]: done

File: _client/src/Dialogs/LoginDialog.cpp (hunk 9 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/LoginDialog.cpp]: done

File: _client/src/Dialogs/RegistrationDialog.cpp (hunk 11 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/RegistrationDialog.cpp]: fixed

File: _client/src/Dialogs/RegistrationDialog.cpp (hunk 10 above)
Comment (MaksGMaks) [reply _client/src/Dialogs/RegistrationDialog.cpp]: done

File: _client/src/UiManager.cpp (hunk 12 above)
Comment (MaksGMaks) [reply _client/src/UiManager.cpp]: restore, this'll be another task

**Pull request merged**